lib/turn_endpoint.ex

```diff
@@ -0,0 +1,624 @@
+    if state.handshake_finished? do
+      {_hsk_state, _hsk_status, hsk_data} = state.handshake
+      event = %Handshake.Event{handshake_data: hsk_data}
+      {{:ok, event: {pad, event}}, state}
+    else
+      {:ok, state}
+    end
+  end
+
+  @impl true
+  def handle_process(
+        Pad.ref(:input, @component_id) = pad,
+        %Membrane.Buffer{payload: payload},
+        %{playback_state: :playing},
+        %{selected_alloc: alloc} = state
+      )
+      when is_pid(alloc) do
+    send(alloc, {:ice_payload, payload})
+    {{:ok, demand: pad}, state}
+  end
+
+  @impl true
+  def handle_process(
+        Pad.ref(:input, @component_id),
+        %Membrane.Buffer{},
+        %{playback_state: :playing},
+        state
+      ) do
+    {{:ok, notify: {:could_not_send_payload, :libnice_not_running}}, state}
+  end
+
+  @impl true
+  def handle_process(_pad, _buffer, %{playback_state: playback_state}, state) do
+    Membrane.Logger.debug("Can't send message in playback state: #{playback_state}. Ignoring.")
+    {:ok, state}
+  end
+
+  @impl true
+  def handle_event(Pad.ref(:input, @component_id) = pad, %Funnel.NewInputEvent{}, _ctx, state) do
+    if state.handshake_finished? do
+      {_hsk_state, _hsk_status, hsk_data} = state.handshake
+      event = {pad, %Handshake.Event{handshake_data: hsk_data}}
+      {{:ok, event: event}, state}
+    else
+      {:ok, state}
+    end
+  end
+
+  @impl true
+  def handle_event(_pad, _event, _ctx, state) do
+    {:ok, state}
+  end
+
+  @impl true
+  def handle_other(:gather_candidates, _ctx, state) do
+    msg = {
+      :new_candidate_full,
+      Utils.generate_fake_ice_candidate(@fake_candidate_addr)
+    }
+
+    {{:ok, notify: msg}, state}
+  end
+
+  @impl true
+  def handle_other({:set_remote_credentials, credentials}, _ctx, state) do
+    [_ice_ufrag, ice_pwd] = String.split(credentials)
+    state = Map.put(state, :remote_ice_pwd, ice_pwd)
+    {:ok, state}
+  end
+
+  @impl true
+  def handle_other(:restart_stream, _ctx, state) do
+    ice_ufrag = Utils.generate_ice_ufrag()
+    ice_pwd = Utils.generate_ice_pwd()
+
+    state = Map.put(state, :local_ice_pwd, ice_pwd)
+    credentials = "#{ice_ufrag} #{ice_pwd}"
+    {{:ok, notify: {:local_credentials, credentials}}, state}
+  end
+
+  @impl true
+  def handle_other(:peer_candidate_gathering_done, _ctx, state) do
+    {:ok, state}
+  end
+
+  @impl true
+  def handle_other({:component_state_ready, @stream_id, @component_id}, ctx, state) do
+    {state, actions} = handle_component_state_ready(ctx, state)
+    {{:ok, actions}, state}
+  end
+
+  @impl true
+  def handle_other({:hsk_finished, @component_id, hsk_data}, ctx, state) do
+    {state, actions} = handle_handshake_finished(hsk_data, ctx, state)
+    {{:ok, actions}, state}
+  end
+
+  @impl true
+  def handle_other({:creating_alloc, alloc_pid}, _ctx, state) do
+    alloc = %{
+      pid: alloc_pid,
+      magic: nil,
+      in_nominated_pair: false,
```

Comment:
There is `in`, instead of `is`, because address of allocation is browser's candidate and flag `in_nominated_pair` says if this allocation is in nominated candidates pair

Reply:
Ach, I see. Could we add one line of comment then?